sync/src/types/mod.rs

```diff
@@ -69,13 +69,60 @@ const SHRINK_THRESHOLD: usize = 300;
 //     If their best known block is still behind when that new timeout is
 //     reached, disconnect.
 
-#[derive(Clone, Debug, Default)]
+#[derive(Clone, Debug)]
 pub struct ChainSyncState {
     pub timeout: u64,
     pub work_header: Option<core::HeaderView>,
     pub total_difficulty: Option<U256>,
     pub sent_getheaders: bool,
-    pub not_sync_until: Option<u64>,
+    headers_sync_state: HeadersSyncState,
+}
+
+impl Default for ChainSyncState {
+    fn default() -> Self {
+        Self {
+            timeout: 0,
+            work_header: None,
+            total_difficulty: None,
+            sent_getheaders: false,
+            headers_sync_state: HeadersSyncState::Initialized,
+        }
+    }
+}
+
+impl ChainSyncState {
+    fn can_sync(&self, now: u64) -> bool {
```

Comment:
The name is confusing. The name `can_start_sync` is a better choice. However, since `HeadersSyncState` already has the function `can_sync`, I'm OK that we continue to use this name in this PR.